Cluster linearization: separate tests from tests-of-tests

sipa opened the pull request:
Part of the cluster mempool project: #30289
 
The current cluster linearization fuzz tests contain two tests which combine testing of production code with testing of the test code itself:
* `clusterlin_search_finder`: establishes the correctness of `SearchCandidateFinder` by comparing against both `SimpleCandidateFinder` and `ExhaustiveCandidateFinder` (which is even more simple than `SimpleCandidateFinder`). If `SimpleCandidateFinder` works correctly, then this comparison with `ExhaustiveCandidateFinder` is redundant. If it isn't, we ought to find that in a test specific to `SimpleCandidateFinder` rather than as a side-effect of testing `SearchCandidateFinder`. Split this functionality out into a new `clusterlin_simple_finder`.
* `clusterlin_linearize`: establishes the correctness of `Linearize` by comparing against both `SimpleLinearize` and literally every valid linearization for the cluster. Again, if `SimpleLinearize` works correctly, then this comparison with all valid linearizations is redundant, and if it isn't we should find it in a test for `SimpleLinearize`. Do so by splitting off that functionality into `clusterlin_simple_linearize`.

After that, a few general improvements to the affected tests are made (comparing with linearizations and subsets read from the fuzz input, plus a performance improvement).

sipa committed:
clusterlin tests: separate testing of Search- and SimpleCandidateFinder

This separates the existing fuzz test into:

* clusterlin_search_finder: establishes SearchCandidateFinder's correctness using the
                            simpler SimpleCandidateFinder.
* clusterlin_simple_finder: establishes SimpleCandidateFinder's correctness using the
                            (even) simpler ExhaustiveCandidateFinder.

rather than trying to do both at once.

sipa committed:
clusterlin tests: separate testing of SimpleLinearize and Linearize

The separates the existing fuzz test into:

* clusterlin_linearize: establishes the correctness of Linearize() using the
                        simpler SimpleLinearize() function.
* clusterlin_simple_linearize: establishes the correctness of SimpleLinearize() by
                        comparing with all valid linearizations computed by
                        std::next_permutation.

rather than combining the first two into a single fuzz test.

ismaelsadeeq reviewed (commented):
Concept ACK

ismaelsadeeq commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -495,11 +494,93 @@ FUZZ_TARGET(clusterlin_ancestor_finder)
+                }
+                // and every union of two ancestor sets.
+                for (auto t1 : todo) {
+                    const auto& a1 = depgraph.Ancestors(t1);
+                    for (auto t2 : todo - a1) {
+                        const auto& a2 = depgraph.Ancestors(t2);
+                        assert(found.feerate >= depgraph.FeeRate((a1 | a2) & todo));
+                    }
```
This can be in the same loop above?

Why not just compare with the result of ancestor finder?

ismaelsadeeq commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -573,22 +650,20 @@ FUZZ_TARGET(clusterlin_search_finder)
-                    assert(exhaustive.feerate == simple.feerate);
+            } else {
+                // If SimpleCandidateFinder couldn't guarantee an optimal result, also test
+                // against every ancestor set ...
+                for (auto t1 : todo) {
+                    const auto& a1 = depgraph.Ancestors(t1);
+                    assert(found.feerate >= depgraph.FeeRate(a1 & todo));
+                }
+                // and every union of two ancestor sets.
+                for (auto t1 : todo) {
+                    const auto& a1 = depgraph.Ancestors(t1);
+                    for (auto t2 : todo - a1) {
+                        const auto& a2 = depgraph.Ancestors(t2);
+                        assert(found.feerate >= depgraph.FeeRate((a1 | a2) & todo));
+                    }
```
We could wrap this in a helper to be dry, as same code is used above.

ismaelsadeeq commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -495,11 +494,93 @@ FUZZ_TARGET(clusterlin_ancestor_finder)
+                // Compare with ExhaustiveCandidateFinder. This quickly gets computationally
+                // expensive for large clusters (O(2^n)), so only do it for sufficiently small ones.
+                auto exhaustive = exh_finder.FindCandidateSet();
+                assert(exhaustive.feerate == found.feerate);
```
Should we also assert that the transactions in the set are the same?

polespinasa replied:
oh right, it's using FeeFrac to compare. Noted!

ismaelsadeeq commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -495,11 +494,93 @@ FUZZ_TARGET(clusterlin_ancestor_finder)
+FUZZ_TARGET(clusterlin_simple_finder)
+{
+    // Verify that SimpleCandidateFinder works as expected by sanity checking the results
+    // and comparing with the results from ExhaustiveCandidateFinder. Note that
```
Maybe indicate that we only compare `SimpleCandidateFinder` with `ExhaustiveCandidateFinder` when we optimally found a subset and the cluster size is small.

ismaelsadeeq reviewed (commented):
Code review bf1b1b09ca4dd18b7848fb1807df8533cf930df7

This is a nice refactor, really improve the cluster linearize test a lot.

ismaelsadeeq commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -74,6 +73,7 @@ class SimpleCandidateFinder
                 // transactions that share ancestry with inc so far (which means only connected
                 // sets will be considered).
                 if (inc_none || inc.Overlaps(m_depgraph.Ancestors(split))) {
+                    --iterations_left;
```
could be in it's own commit?

sipa committed:
clusterlin tests: count SimpleCandidateFinder iterations better

Only count the number of actual new subsets added. If the queue contains
a work item that completely covers a component, no transaction can be added
to it without creating a disconnected component. In this case, also don't
count it as an iteration.

With this, the number of iterations performed by SimpleCandidateFinder is
bounded by the number of distinct connected topologically-valid subsets of
the cluster.

sipa committed:
clusterlin tests: optimize clusterlin_simple_linearize

Whenever a non-topological permutation is encountered, fast forward to the
last permutation with the same non-topological prefix, skipping over
potentially many permutations that are non-topological for the same reason.

With that, increase the checking of all permutations to clusters of size 8
instead of 7.

sipa committed:
clusterlin tests: compare with fuzz-provided topological sets

sipa committed:
clusterlin tests: compare with fuzz-provided linearizations

ismaelsadeeq reviewed (commented):
Code review ACK 0cae0c487cb8b32b96f888a31e818f2a90f848cc

ismaelsadeeq commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -559,6 +559,10 @@ FUZZ_TARGET(clusterlin_simple_finder)
             }
+
+            // Compare with a topological set read from the fuzz input.
+            auto read_topo = ReadTopologicalSet(depgraph, todo, reader);
```
We can cache the subset and then reuse it here and when finding topological subset to be deleted.

ismaelsadeeq commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -659,6 +663,10 @@ FUZZ_TARGET(clusterlin_search_finder)
             auto anc = anc_finder.FindCandidateSet();
             assert(found.feerate >= anc.feerate);
+
+            // Compare with a topological set read from the fuzz input.
```
Same  comment applies here as well.

ismaelsadeeq replied:
I see your point now.
both comments can be resolved.

ismaelsadeeq reviewed (commented):
I reviewed the latest changes in the last push 

Code review 8eb70e3ca467ac66bc1f43e3e0b3338d63806ee5

Changes since my last review were

1. Addressed my previous review comments.

2. An optimization was introduced in the  `clusterlin_simple_linearize` fuzz test that skips to the last non-topological permutation.
For example, given transactions `[A, B, C, D, E]`.
if A-B-C are a valid topological ancestor set, and the first permutation results in [A, E, B, C, D], which is not topologically valid, we can skip to [A, E, D, C, B]. 
This optimization saves iterations, allowing us to increase the maximum dependency graph we can check to 8.

3. Assert that the found subsets are compared against the first valid topological set, verifying that we achieve better results when optimal.

4. Compared the optimal linearization to the default chunking of the dependency graph.

Overall, the changes look good to me. I left some comments.
I also have a single question regarding the process of marking topological set as done.

ismaelsadeeq commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -495,11 +494,94 @@ FUZZ_TARGET(clusterlin_ancestor_finder)
+            }
+        }
+
+        // Find a topologically valid subset of transactions to remove from the graph.
+        auto del_set = ReadTopologicalSet(depgraph, todo, reader);
+        // If we did not find anything, use found itself, because we should remove something.
+        if (del_set.None()) del_set = found.transactions;
+        todo -= del_set;
```
Thanks for clarifying, initially I thought the aim is to remove the subset found by the search continuously until there are no transactions left in `m_todo`.

Maybe it will be helpful to document the point you gave as comment?

sipa committed:
clusterlin tests: support non-empty ReadTopologicalSubset()

In several call sites for ReadTopologicalSubset, a non-empty result is
expected, necessitating a special case at the call site for empty results.

Fix this by adding a bool non_empty argument, which does this special
casing (more efficiently) inside ReadTopologicalSubset itself.

ismaelsadeeq reviewed (approved):
Code review ACK 46b6a67eccb411903deb09af3301c2499d2c0217

All comments were addressed.

ismaelsadeeq commented:
re-ACK 30f30bbcaae67d725360c3e7b83207904e1a9b19 

The recent push was rebase on master after https://github.com/bitcoin/bitcoin/pull/30857

marcofleon commented:
```
echo "oz0QFv7///////8HPRCGEBAQEBA7Ozs7OwYGBgYGd3d3d3d3d5sAAAAAAAAAd3d3d3cGBgYGBgb/BgYGBgYGBgYGCAgICAgICAgICAgICAgICAgICAgICAgIBgYGBgYGBgYGBgY7OzsQEBAQEBAQEBAQEBAQLT09PRA9PT09PRAQEBAQEC0QEBAQLT09PRA9PT09EA==" | base64 --decode > clusterlin_linearize.crash
FUZZ=clusterlin_linearize ./clusterlinfuzzbuild/src/test/fuzz/fuzz clusterlin_linearize.crash
src/test/fuzz/cluster_linearize.cpp:1069: void clusterlin_linearize_fuzz_target(FuzzBufferType): Assertion `cmp_read >= 0' failed.
```
Just did some initial fuzzing for now, but I'm looking to do a more thorough review of this soon.

sipa committed:
clusterlin tests: verify that chunks are minimal

ismaelsadeeq reviewed (commented):
reACK d2fc10b64505512a70bed05e7bb21d76c8f748cd

With some comments, even though some things added here are deleted in #32545
I think this is quite useful and should get in before SFL.

sipa committed:
clusterlin tests: make SimpleCandidateFinder always find connected

Make a small change to guarantee that SimpleCandidateFinder only ever returns
connected solutions, even when non-optimal. Then test this property.

ismaelsadeeq reviewed (approved):
reACK 30c699dea39157631515d867c65a03987bd46cce

sipa committed:
clusterlin: abstract try-permutations into ExhaustiveLinearize function

Rather than this exhaustive linearization check happening inline inside
clusterlin_simple_linearize, abstract it out into a Linearize()-like
function for clarity.

Note that this isn't exactly a refactor, because the old code would compare the
found linearization against all (valid) permutations, while the new code instead
first computes the best linearization from all valid permutations, and then
compares it with the found one.

sipa committed:
clusterlin: add big comment explaning the relation between tests

marcofleon reviewed (commented):
code review ACK 9877e275b95400e46f336bc34ce41ed68ce43683

Previously, the test helpers (`SimpleCandidateFinder` and `SimpleLinearize`) were being checked for correctness within the same fuzz targets that are testing the production cluster linearization. Separating out these checks into their own targets reduces redundancy and makes the hierarchy of all these tests easier to follow. 

I ran `clusterlin_linearize`, `clusterlin_search_finder`, `clusterlin_simple_finder`, and `clusterlin_simple_linearize` for ~50 CPU hours each, no issues.

The added diagram comment at the beginning is helpful. It also looks a lot cleaner in #32545 😎

ismaelsadeeq commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -17,6 +17,67 @@
+/*
+ * The tests in this file primarily test the candidate finder classes and linearization algorithms.
+ *
+ *   <----: the right side is tested in the named test by comparison with the left side
```
In "clusterlin: add big comment explaning the relation between tests
" 9877e275b95400e46f336bc34ce41ed68ce43683

I like the new comment addition, just the first description is a bit unclear because the line is diagonal

Maybe this suggestion will be clearer?
```diff
diff --git a/src/test/fuzz/cluster_linearize.cpp b/src/test/fuzz/cluster_linearize.cpp
index 7d0fb5fbec0..8820e3b3eaa 100644
--- a/src/test/fuzz/cluster_linearize.cpp
+++ b/src/test/fuzz/cluster_linearize.cpp
@@ -18,10 +18,11 @@
 #include <utility>
 
 /*
- * The tests in this file primarily test the candidate finder classes and linearization algorithms.
+ * The tests in this file primarily cover the candidate finder classes and linearization algorithms.
  *
- *   <----: the right side is tested in the named test by comparison with the left side
- *   <<---: right side is implemented using left side
+ *   <----: An implementation (at the start of the line --) is tested in the test marked with *
+ *          by comparison with other implementations (at the end of the line ->).
+ *   <<---: The right side is implemented using the left side.
  *
  *   +-----------------------+
  *   | SearchCandidateFinder | <<---------------------\
@@ -38,7 +39,7 @@
  *     |                            |                    |        ||
  *     |                            |                    |        vv  TEST CODE
  *     |                            |                    |
- *     |-clusterlin_search_finder   |                    |-clusterlin_linearize
+ *     |-clusterlin_search_finder*  |                    |-clusterlin_linearize*
  *     |                            |                    |
  *     v                            |                    v
  *   +-----------------------+      |               +-----------------+
@@ -47,7 +48,7 @@
  *                  |               |                    |
  *                  +---------------/                    |
  *                  |                                    |
- *                  |-clusterlin_simple_finder           |-clusterlin_simple_linearize
+ *                  |-clusterlin_simple_finder*          |-clusterlin_simple_linearize*
  *                  v                                    v
  *   +---------------------------+                  +---------------------+
  *   | ExhaustiveCandidateFinder |                  | ExhaustiveLinearize |
abubakarismail@Abubakars-MacBook-Pro ~/D/W/b/bitcoin ((9877e275))> 
```

ismaelsadeeq reviewed (commented):
re-ACK 9877e275b95400e46f336bc34ce41ed68ce43683

ismaelsadeeq reviewed (approved):
re-ACK d7fca5c171f450621112457ea6f7c99b2e21d354

monlovesmango reviewed (commented):
ACK d7fca5c171f450621112457ea6f7c99b2e21d354

The separation of concerns in the cluster linearization fuzz tests makes sense, namely separating testing of production of code from testing of test code. The small added efficiencies and comparisons are nice and the diagram documenting all the relationships is extremely helpful in clarifying how all the parts tie together.

Unrelated to this specific PR, but in the `clusterlin_postlinearize_tree` fuzz target, `post_post_linearization` isn't actually getting used in the `PostLinearize` and `SanityCheck` calls. Both calls are using `post_linearize` as the second parameter instead of `post_post_linearize`. Does this need to be corrected? See below for current code snippet.
```cpp
    // Verify that post-linearizing again does not change the diagram. The result must be identical
    // as post_linearization ought to be optimal already with a tree-structured graph.
    auto post_post_linearization = post_linearization;
    PostLinearize(depgraph_tree, post_linearization);
    SanityCheck(depgraph_tree, post_linearization);
    auto post_post_chunking = ChunkLinearization(depgraph_tree, post_post_linearization);
    auto cmp_post = CompareChunks(post_post_chunking, post_chunking);
    assert(cmp_post == 0);
```

Another question unrelated to this PR, in `clusterlin_simple_linearize` before calling `assert(optimal)` why does it check `n` (`= depgraph.TxCount()`) to be less than or equal to 63 but in `clusterlin_linearize` it checks `n` to be less than or equal to 19? I had thought that maybe the check against 63 was due to the use of `uint64_t` type but I couldn't figure out where the 19 comes from. I also fuzzed with the 19 value changed to 63 for a while and didn't get any failures.

ismaelsadeeq commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -185,20 +296,34 @@ void MakeConnected(DepGraph<BS>& depgraph)
-SetType ReadTopologicalSet(const DepGraph<SetType>& depgraph, const SetType& todo, SpanReader& reader)
+SetType ReadTopologicalSet(const DepGraph<SetType>& depgraph, const SetType& todo, SpanReader& reader, bool non_empty)
 {
+    // Read a bitmask from the fuzzing input. Add 1 if non_empty, so the mask is definitely not
+    // zero in that case.
     uint64_t mask{0};
     try {
         reader >> VARINT(mask);
     } catch(const std::ios_base::failure&) {}
+    mask += non_empty;
```
Maybe not because comment above explicitly mentioned add, but that can be modified to match |.

see https://github.com/ismaelsadeeq/bitcoin/commit/95ea4e737612c094dfcb6414f8005c79c6a34551

ajtowns replied:
`if (!mask) mask += nonzero;` would be simpler, no? Or would that bias the fuzzer in undesirable ways somehow? Could perhaps do `mask += (!mask & nonzero)` to avoid an explicit branch.
